enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,1090 @@
+	// The current default value is "RollingUpdate".
+	// +kubebuilder:default:="RollingUpdate"
+	// +kubebuilder:validation:Enum:="RollingUpdate";"OnDelete"
+	// +optional
+	Type ControlPlaneMachineSetStrategyType `json:"type,omitempty"`
+
+	// This is left as a struct to allow future rolling update
+	// strategy configuration to be added later.
+}
+
+// ControlPlaneMachineSetStrategyType is an enumeration of different update strategies
+// for the Control Plane Machines.
+type ControlPlaneMachineSetStrategyType string
+
+const (
+	// RollingUpdate is the default update strategy type for a
+	// ControlPlaneMachineSet. This will cause the ControlPlaneMachineSet to
+	// first create a new Machine and wait for this to be Ready
+	// before removing the Machine chosen for replacement.
+	RollingUpdate ControlPlaneMachineSetStrategyType = "RollingUpdate"
+
+	// Recreate causes the ControlPlaneMachineSet controller to first
+	// remove a ControlPlaneMachine before creating its
+	// replacement. This allows for scenarios with limited capacity
+	// such as baremetal environments where additional capacity to
+	// perform rolling updates is not available.
+	Recreate ControlPlaneMachineSetStrategyType = "Recreate"
+
+	// OnDelete causes the ControlPlaneMachineSet to only replace a
+	// Machine once it has been marked for deletion. This strategy
+	// makes the rollout of updated specifications into a manual
+	// process. This allows users to test new configuration on
+	// a single Machine without forcing the rollout of all of their
+	// Control Plane Machines.
+	OnDelete ControlPlaneMachineSetStrategyType = "OnDelete"
+)
+
+// FailureDomain represents the different configurations required to spread Machines
+// across failure domains on different platforms.
+// +union
+type FailureDomains struct {
+	// Platform identifies the platform for which the FailureDomain represents
+	// +unionDiscriminator
+	// +optional
+	Platform configv1.PlatformType `json:"platform,omitempty"`
+
+	// AWS configures failure domain information for the AWS platform
+	// +optional
+	AWS *[]AWSFailureDomain `json:"aws,omitempty"`
+
+	// Azure configures failure domain information for the Azure platform
+	// +optional
+	Azure *[]AzureFailureDomain `json:"azure,omitempty"`
+
+	// GCP configures failure domain information for the GCP platform
+	// +optional
+	GCP *[]GCPFailureDomain `json:"gcp,omitempty"`
+
+	// OpenStack configures failure domain information for the OpenStack platform
+	// +optional
+	OpenStack *[]OpenStackFailureDomain `json:"openstack,omitempty"`
+}
+
+// AWSFailureDomain configures failure domain information for the AWS platform
+type AWSFailureDomain struct {
+	// Subnet is a reference to the subnet to use for this instance
+	// +optional
+	Subnet AWSResourceReference `json:"subnet,omitempty"`
+
+	// Placement configures the placement information for this instance
+	// +optional
+	Placement AWSFailureDomainPlacement `json:"placement,omitempty"`
```

Comment:
Do we need the placement, as it is redundant with the subnet? The subnet is already associated with a specific availability zone.

Reply:
The subnet is optional in Machine API, it has the ability to lookup a default subnet based on the availability zone provided if the subnet is not provided. As such, we need to maintain this behaviour between this API and the main AWS providerSpec API

Reply:
OK. It was not obvious to me from the comments that the subnet is optional when specifying placements for AWS. With the union, it is difficult to ascertain which fields are actually optional for a given platform. I recommend that we add some more text to the description of `subnet` indicating that a user can be omit it to have the operator select the appropriate subnet for the user (and perhaps the limitations of that).

Reply:
The kubeubilder tags for optional and required should be accurate, for example there's a tag on this struct that indicates you must set at least one of the values within it, but both are optional, which makes the openapi error at you if you provide an empty spec, but if either field is set, it will accept it.

I will expand the subnet comment though as you're right, it's a little light in detail at the moment